File: frontend/src/api/modules/system.ts
import { apiClient } from '@/api/client'
import type { ApiEnvelope } from '@/api/types'

export interface ProviderRow {
  id: string
  name: string
  status: string
  providerType?: string
  provider_type?: string
  baseUrl?: string
  base_url?: string
  defaultModel?: string
  default_model?: string
  enabled?: boolean
  apiKeyConfigured?: boolean
  api_key_configured?: boolean
  apiKeyMasked?: string
  api_key_masked?: string
}

export type ProviderType = 'openai_compatible' | 'anthropic' | 'gemini'

export interface QuotaPolicyRow {
  id: string
  scope: string
  dailyCostLimit: number
}

export interface LLMSettings {
  providerName?: string
  provider_name?: string
  baseUrl?: string
  base_url?: string
  model: string
  apiKeyConfigured?: boolean
  api_key_configured?: boolean
}

export interface SourceBuildAgentSettings {
  enabled: boolean
  providerConfigured?: boolean
  provider_configured?: boolean
}

export interface InteractiveBrowserSettings {
  enabled: boolean
  automaticEnabled: boolean
  maxSessions: number
  sessionTimeoutSeconds: number
}

export interface InteractiveBrowserSettingsInput {
  enabled: boolean
  automaticEnabled: boolean
  maxSessions: number
  sessionTimeoutSeconds: number
}

interface InteractiveBrowserSettingsResponse {
  enabled: boolean
  automatic_enabled: boolean
  max_sessions: number
  session_timeout_seconds: number
}

export interface ProviderConfigurationInput {
  name: string
  providerType?: ProviderType
  baseUrl: string
  apiKey: string
  defaultModel: string
  enabled: boolean
}

export interface ProviderRouteEntry {
  id?: string
  providerAccountId?: string
  provider_account_id?: string
  providerName?: string
  provider_name?: string
  model: string
  priority?: number
  enabled: boolean
}

export interface ProviderRoute {
  group: string
  entries: ProviderRouteEntry[]
}

export interface NovelUsageMetrics {
  requests: number
  cache_hits: number
  cache_misses: number
  input_tokens: number
  output_tokens: number
  cost: number
  evidence_count: number
  providers: string[]
  models: Record<string, string[]>
}

export interface NovelSettings {
  vector_backend?: string
  endpoint?: string
  collection_prefix?: string
  dimension?: number
  embedding_model?: string
  batch_size?: number
  threshold?: number
  concurrency?: number
  retries?: number
  cache_ttl?: number
  enabled_tools?: string[]
  chapter_size?: number
  index_policy?: string
  cost_budget_daily?: number
  cost_budget_per_request?: number
  credential_configured?: boolean
  credential_masked?: string
  route_groups?: Record<string, Array<{ provider: string; model: string; enabled: boolean }>>
  models?: Record<string, string[]>
  metrics?: NovelUsageMetrics
}

export interface SettingsSection<T extends Record<string, unknown> = Record<string, unknown>> {
  domain: string
  tab: string
  value: T
  version: string | null
  updatedAt?: string | null
  updated_at?: string | null
}

export function listProviders(): Promise<ApiEnvelope<ProviderRow[]>> {
  return apiClient.get<ProviderRow[]>('/system/providers')
}

export function createProvider(payload: ProviderConfigurationInput): Promise<ApiEnvelope<ProviderRow>> {
  return apiClient.post<ProviderRow>('/system/providers', {
    name: payload.name,
    ...(payload.providerType ? { provider_type: payload.providerType } : {}),
    base_url: payload.baseUrl,
    api_key: payload.apiKey,
    default_model: payload.defaultModel,
    enabled: payload.enabled,
  })
}

export function updateProvider(providerId: string, payload: ProviderConfigurationInput): Promise<ApiEnvelope<ProviderRow>> {
  return apiClient.put<ProviderRow>(`/system/providers/${providerId}`, {
    name: payload.name,
    ...(payload.providerType ? { provider_type: payload.providerType } : {}),
    base_url: payload.baseUrl,
    api_key: payload.apiKey,
    default_model: payload.defaultModel,
    enabled: payload.enabled,
  })
}

export function discoverProviderModels(providerId: string): Promise<ApiEnvelope<string[]>> {
  return apiClient.post<string[]>(`/system/providers/${providerId}/models`)
}

export function getProviderRoute(group: string): Promise<ApiEnvelope<ProviderRoute>> {
  return apiClient.get<ProviderRoute>(`/system/provider-routes/${group}`)
}

export function updateProviderRoute(
  group: string,
  payload: { entries: Array<{ providerAccountId: string; model: string; enabled: boolean }> },
): Promise<ApiEnvelope<ProviderRoute>> {
  return apiClient.put<ProviderRoute>(`/system/provider-routes/${group}`, {
    entries: payload.entries.map((entry) => ({
      provider_account_id: entry.providerAccountId,
      model: entry.model,
      enabled: entry.enabled,
    })),
  })
}

export function listQuotaPolicies(): Promise<ApiEnvelope<QuotaPolicyRow[]>> {
  return apiClient.get<QuotaPolicyRow[]>('/system/quotas')
}

export function getLLMSettings(): Promise<ApiEnvelope<LLMSettings>> {
  return apiClient.get<LLMSettings>('/system/llm-settings')
}

export function updateLLMSettings(payload: {
  providerName: string
  baseUrl: string
  apiKey: string
  model: string
}): Promise<ApiEnvelope<LLMSettings>> {
  return apiClient.put<LLMSettings>('/system/llm-settings', {
    provider_name: payload.providerName,
    base_url: payload.baseUrl,
    api_key: payload.apiKey,
    model: payload.model,
  })
}

export function getSourceBuildAgentSettings(): Promise<ApiEnvelope<SourceBuildAgentSettings>> {
  return apiClient.get<SourceBuildAgentSettings>('/system/source-build-agent-settings')
}

export function updateSourceBuildAgentSettings(payload: {
  enabled: boolean
}): Promise<ApiEnvelope<SourceBuildAgentSettings>> {
  return apiClient.put<SourceBuildAgentSettings>('/system/source-build-agent-settings', payload)
}

export function getNovelSettings(): Promise<ApiEnvelope<NovelSettings>> {
  return apiClient.get<NovelSettings>('/system/novel-settings')
}

export function updateNovelSettings(payload: Partial<NovelSettings>): Promise<ApiEnvelope<NovelSettings>> {
  return apiClient.put<NovelSettings>('/system/novel-settings', payload)
}

export function getSettingsSection<T extends Record<string, unknown> = Record<string, unknown>>(
  domain: string,
  tab: string,
): Promise<ApiEnvelope<SettingsSection<T>>> {
  return apiClient.get<SettingsSection<T>>(`/system/settings/${domain}/${tab}`)
}

export function saveSettingsSection<T extends Record<string, unknown> = Record<string, unknown>>(
  domain: string,
  tab: string,
  value: T,
  expectedVersion: string | null,
): Promise<ApiEnvelope<SettingsSection<T>>> {
  return apiClient.put<SettingsSection<T>>(`/system/settings/${domain}/${tab}`, {
    value,
    expected_version: expectedVersion,
  })
}

function mapInteractiveBrowserSettings(settings: InteractiveBrowserSettingsResponse): InteractiveBrowserSettings {
  return {
    enabled: settings.enabled,
    automaticEnabled: settings.automatic_enabled,
    maxSessions: settings.max_sessions,
    sessionTimeoutSeconds: settings.session_timeout_seconds,
  }
}

export async function getInteractiveBrowserSettings(): Promise<ApiEnvelope<InteractiveBrowserSettings>> {
  const response = await apiClient.get<InteractiveBrowserSettingsResponse>('/system/interactive-browser-settings')
  return {
    ...response,
    data: mapInteractiveBrowserSettings(response.data),
  }
}

export async function updateInteractiveBrowserSettings(
  payload: InteractiveBrowserSettingsInput,
): Promise<ApiEnvelope<InteractiveBrowserSettings>> {
  const response = await apiClient.put<InteractiveBrowserSettingsResponse>('/system/interactive-browser-settings', {
    enabled: payload.enabled,
    automatic_enabled: payload.automaticEnabled,
    max_sessions: payload.maxSessions,
    session_timeout_seconds: payload.sessionTimeoutSeconds,
  })
  return {
    ...response,
    data: mapInteractiveBrowserSettings(response.data),
  }
}

File: backend/app/interfaces/http/system.py
from datetime import datetime
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field

from app.core.config import settings
from app.core.permissions import Permission
from app.core.response import ok
from app.application.ports.provider import provider_http_status
from app.domain.repositories.system_settings_repo import ConcurrentSettingsUpdateError
from app.infrastructure.persistence.factory import (
    build_provider_platform_service,
    build_system_settings_service,
    build_vector_store,
)
from app.interfaces.http.deps import require_permission


router = APIRouter()


class LLMSettingsRequest(BaseModel):
    provider_name: str = Field(min_length=1, max_length=120)
    base_url: str = Field(min_length=1, max_length=500)
    api_key: str = Field(default="", max_length=2000)
    model: str = Field(min_length=1, max_length=120)


class SourceBuildAgentSettingsRequest(BaseModel):
    enabled: bool


class InteractiveBrowserSettingsRequest(BaseModel):
    enabled: bool
    automatic_enabled: bool = True
    max_sessions: int = Field(default=1, ge=1, le=3)
    session_timeout_seconds: int = Field(default=300, ge=60, le=600)


class SettingsSectionRequest(BaseModel):
    value: dict[str, object]
    expected_version: str | None = None


class ProviderRequest(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    provider_type: str | None = Field(default=None, pattern="^(openai_compatible|anthropic|gemini)$")
    base_url: str = Field(min_length=1, max_length=500)
    api_key: str = Field(default="", max_length=2000)
    default_model: str = Field(default="", max_length=120)
    enabled: bool = True
    activation_at: datetime | None = None


class ProviderRouteEntryRequest(BaseModel):
    provider_account_id: str = Field(min_length=1, max_length=100)
    model: str = Field(min_length=1, max_length=120)
    enabled: bool = True


class ProviderRouteRequest(BaseModel):
    entries: list[ProviderRouteEntryRequest] = Field(min_length=1, max_length=20)


class NovelSettingsRequest(BaseModel):
    vector_backend: str | None = Field(default=None, min_length=1, max_length=20)
    backend: str | None = Field(default=None, min_length=1, max_length=20)
    endpoint: str | None = Field(default=None, max_length=500)
    api_key: str | None = Field(default=None, max_length=2000)
    collection_prefix: str | None = Field(default=None, min_length=1, max_length=80)
    dimension: int | None = Field(default=None, ge=0, le=65536)
    embedding_model: str | None = Field(default=None, max_length=200)
    batch_size: int | None = Field(default=None, ge=1, le=256)
    threshold: float | None = Field(default=None, ge=0, le=1)
    concurrency: int | None = Field(default=None, ge=1, le=64)
    retries: int | None = Field(default=None, ge=0, le=10)
    cache_ttl: int | None = Field(default=None, ge=0)
    enabled_tools: list[str] | None = Field(default=None, max_length=3)
    chapter_size: int | None = Field(default=None, ge=1000, le=100000)
    index_policy: str | None = Field(default=None, pattern="^(incremental|full)$")
    cost_budget_daily: float | None = Field(default=None, ge=0)
    cost_budget_per_request: float | None = Field(default=None, ge=0)


def _system_response(message: str, data):
    return ok(data=data, message=message, meta={})


@router.get("/settings/{domain}/{tab}")
async def get_settings_section(
    domain: str,
    tab: str,
    _=Depends(require_permission(Permission.SYSTEM_SETTINGS_MANAGE)),
):
    try:
        data = build_system_settings_service().get_section(domain, tab)
    except ValueError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    return _system_response("settings section loaded", data)


@router.put("/settings/{domain}/{tab}")
async def save_settings_section(
    domain: str,
    tab: str,
    payload: SettingsSectionRequest,
    _=Depends(require_permission(Permission.SYSTEM_SETTINGS_MANAGE)),
):
    try:
        data = build_system_settings_service().save_section(
            domain,
            tab,
            payload.value,
            expected_version=payload.expected_version,
        )
    except ConcurrentSettingsUpdateError as exc:
        raise HTTPException(status_code=409, detail="settings section was updated by another request") from exc
    except ValueError as exc:
        status_code = 404 if str(exc).startswith("Unknown settings section:") else 422
        raise HTTPException(status_code=status_code, detail=str(exc)) from exc
    return _system_response("settings section saved", data)


@router.get("/providers")
async def list_providers(_=Depends(require_permission(Permission.SYSTEM_SETTINGS_MANAGE))):
    service = build_provider_platform_service()
    data = service.list_provider_accounts()
    return ok(data=data, message="providers listed", meta={"total": len(data)})


@router.post("/providers")
async def create_provider(
    payload: ProviderRequest,
    _=Depends(require_permission(Permission.SYSTEM_SETTINGS_MANAGE)),
):
    service = build_provider_platform_service()
    try:
        data = service.save_provider(**payload.model_dump())
    except ValueError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc
    return _system_response("provider saved", data)


@router.put("/providers/{provider_id}")
async def update_provider(
    provider_id: str,
    payload: ProviderRequest,
    _=Depends(require_permission(Permission.SYSTEM_SETTINGS_MANAGE)),
):
    service = build_provider_platform_service()
    try:
        data = service.save_provider(provider_id=provider_id, **payload.model_dump())
    except LookupError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc
    return _system_response("provider saved", data)


@router.post("/providers/{provider_id}/models")
async def discover_provider_models(
    provider_id: str,
    _=Depends(require_permission(Permission.SYSTEM_SETTINGS_MANAGE)),
):
    service = build_provider_platform_service()
    try:
        data = await service.discover_models(provider_id)
    except LookupError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc
    except Exception as exc:
        status = provider_http_status(exc)
        if status in {401, 403}:
            detail = "Provider authentication failed; update the API key"
            status_code = 422
        elif status is not None:
            detail = f"Provider model discovery failed (HTTP {status})"
            status_code = 502
        else:
            detail = "Provider model discovery is unavailable"
            status_code = 502
        raise HTTPException(status_code=status_code, detail=detail) from exc
    return _system_response("provider models listed", data)


@router.get("/provider-routes/{provider_group}")
async def get_provider_route(
    provider_group: str,
    _=Depends(require_permission(Permission.SYSTEM_SETTINGS_MANAGE)),
):
    try:
        data = build_provider_platform_service().get_routes(provider_group)
    except ValueError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc
    return _system_response("provider route loaded", data)


@router.put("/provider-routes/{provider_group}")
async def update_provider_route(
    provider_group: str,
    payload: ProviderRouteRequest,
    _=Depends(require_permission(Permission.SYSTEM_SETTINGS_MANAGE)),
):
    try:
        data = build_provider_platform_service().replace_routes(
            provider_group,
            [entry.model_dump() for entry in payload.entries],
        )
    except LookupError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc
    return _system_response("provider route saved", data)


@router.get("/llm-settings")
async def get_llm_settings(_=Depends(require_permission(Permission.SYSTEM_SETTINGS_MANAGE))):
    service = build_provider_platform_service()
    data = service.get_llm_settings(
        default_provider_name=settings.LLM_PROVIDER_NAME,
        default_model=settings.LLM_MODEL,
    )
    return ok(data=data, message="llm settings loaded", meta={})


@router.put("/llm-settings")
async def update_llm_settings(
    payload: LLMSettingsRequest,
    _=Depends(require_permission(Permission.SYSTEM_SETTINGS_MANAGE)),
):
    service = build_provider_platform_service()
    data = service.save_llm_settings(
        provider_name=payload.provider_name,
        base_url=payload.base_url,
        api_key=payload.api_key,
        model=payload.model,
    )
    return ok(data=data, message="llm settings saved", meta={})


@router.get("/source-build-agent-settings")
async def get_source_build_agent_settings(
    _=Depends(require_permission(Permission.SYSTEM_SETTINGS_MANAGE)),
):
    data = build_system_settings_service().get_source_build_agent_settings()
    return ok(data=data, message="source build agent settings loaded", meta={})


@router.put("/source-build-agent-settings")
async def update_source_build_agent_settings(
    payload: SourceBuildAgentSettingsRequest,
    _=Depends(require_permission(Permission.SYSTEM_SETTINGS_MANAGE)),
):
    data = build_system_settings_service().set_source_build_agent_enabled(payload.enabled)
    return ok(data=data, message="source build agent settings saved", meta={})


@router.get('/interactive-browser-settings')
async def get_interactive_browser_settings(
    _=Depends(require_permission(Permission.SYSTEM_SETTINGS_MANAGE)),
):
    return _system_response(
        'interactive browser settings loaded',
        build_system_settings_service().get_interactive_browser_settings(),
    )


@router.put('/interactive-browser-settings')
async def update_interactive_browser_settings(
    payload: InteractiveBrowserSettingsRequest,
    _=Depends(require_permission(Permission.SYSTEM_SETTINGS_MANAGE)),
):
    return _system_response(
        'interactive browser settings saved',
        build_system_settings_service().set_interactive_browser_settings(**payload.model_dump()),
    )


@router.get("/novel-settings")
async def get_novel_settings(
    _=Depends(require_permission(Permission.SYSTEM_SETTINGS_MANAGE)),
):
    data = build_system_settings_service().get_novel_settings()
    return _system_response("novel settings loaded", data)


@router.put("/novel-settings")
async def update_novel_settings(
    payload: NovelSettingsRequest,
    _=Depends(require_permission(Permission.SYSTEM_SETTINGS_MANAGE)),
):
    try:
        data = build_system_settings_service().set_novel_settings(payload.model_dump(exclude_unset=True))
    except ValueError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc
    return _system_response("novel settings saved", data)


@router.post("/novel-settings/test-vector-store")
async def test_novel_vector_store(
    _=Depends(require_permission(Permission.SYSTEM_SETTINGS_MANAGE)),
):
    service = build_system_settings_service()
    configured = service.get_novel_settings()
    store = build_vector_store()
    try:
        data = await store.health()
    except Exception as exc:
        data = {
            "enabled": False,
            "backend": configured.get("vector_backend", "disabled"),
            "error": str(exc)[:200],
        }
    finally:
        close = getattr(store, "aclose", None)
        if callable(close):
            result = close()
            if hasattr(result, "__await__"):
                await result
    return _system_response("novel vector store tested", data)


@router.get("/quotas")
async def list_quota_policies(_=Depends(require_permission(Permission.SYSTEM_SETTINGS_MANAGE))):
    service = build_provider_platform_service()
    data = service.list_quota_policies()
    return ok(data=data, message="quota policies listed", meta={"total": len(data)})
